### PR TITLE
Add mock template locator

### PR DESCRIPTION
**Pull request by wli3**
Check in the code, so it will be published to the feed.
dotnet/templating can start to consume it. Once the actual
implementation is done, we can rename the package. The only thing
dotnet/templating need to do is to change the package reference name
then.

### CODE REVIEW

**Comment by wli3** [comment]
@dsplaisted seems OOF today. However, I promised @grinrag and @DonJoseLuis to finish it today.

@marcpopMSFT could you review this. Once this is merged. A package `Microsoft.Internal.DotNet.MockTemplateLocator` will be published to the dev feed. The consumer is internal. So there is no harm. But we can start the integration.

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (joeloff) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:11]: So this is just a placeholder to unblock partner teams?

Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:11]: Correct

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (joeloff) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:50]: Can this ever be null?

Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:50]: no. it is an internal class. The call site won't pass null

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (sfoslund) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:19]: nit: mockMock?

File: src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs
Comment (sfoslund) [inline src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs:32]: Is there a way to have more extensive assertions here?

**Review by sfoslund — APPROVED** [review]

File: src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs
Comment (wli3) [reply src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs:32]: Cannot do that reliably. I just lists out the templates in stage2 template folder, and we don't really control what goes in there (inserted from dotnet/templating). Plus, the proper implementation will only read the manifest file instead of probing the disk. This is good enough.

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (marcpopMSFT) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:19]: Sometimes we put _ in our environment variable names and sometimes we don't.  Is there a reason we choose one or the other?

File: src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs
Comment (marcpopMSFT) [inline src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs:27]: If the directory doesn't exist, won't GetDirectories() return an empty array and then trigger an exception when you look at the [0] item

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:19]: It is for testing. A temporary implementation. We won't need it when it has the proper implementation

File: src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs
Comment (wli3) [reply src/Tests/Microsoft.DotNet.TemplateLocator.Tests/GivenAnMSBuildSdkResolver.cs:27]: That's fine. The only user see it is the dev working on _dotnet/templating_. And they can react.

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (marcpopMSFT) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:34]: Can GetPackagesV2 return null?

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (marcpopMSFT) [inline src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:40]: Should this be centralized eventually?

Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:40]: This is a place holder implementation. Template engine currently don't need this. But it is part of the spec we decided on (for VS integration later)

File: src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs
Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs:34]: Hmm actually it would. https://github.com/NuGet/NuGet.Client/blob/784274cc9b2002ca73055e8134064261118d8a01/src/NuGet.Core/NuGet.Protocol/Utility/LocalFolderUtility.cs#L194

I will throw an exception when it is null

Comment (wli3) [reply src/Microsoft.DotNet.TemplateLocator/TemplateLocator.cs]: Just made it return empty array when null

**Comment by wli3** [comment]
@marcpopMSFT just replied all your comments

**Review by joeloff — APPROVED** [review]

**Comment by wli3** [comment]
@grinrag this is merged. Once the official build is done. It should show up in feed: https://pkgs.dev.azure.com/dnceng/internal/_packaging/dotnet5-internal-transport/nuget/v3/index.json

**Pull request merged**
